### PR TITLE
[SPARK-19799][SQL] Support WITH clause in subqueries

### PR DESCRIPTION
**Pull request by peter-toth**
## What changes were proposed in this pull request?

This PR  adds support of `WITH` clause within a subquery so this query becomes valid:
  ```
  SELECT max(c) FROM (
    WITH t AS (SELECT 1 AS c)
    SELECT * FROM t
  )
 ```

## How was this patch tested?

Added new UTs.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
cc @gatorsmile since this is a part of PostgreSQL feature parity.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 1]
```diff
@@ -412,7 +412,7 @@ queryPrimary
     | fromStatement                                                         #fromStmt
     | TABLE multipartIdentifier                                             #table
     | inlineTable                                                           #inlineTableDefault1
-    | '(' queryNoWith  ')'                                                  #subquery
+    | '(' query  ')'                                                        #subquery
```
Comment (maropu) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: nit: remove an unnecessary single space.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -212,31 +212,96 @@ class Analyzer(
   )
 
   /**
-   * Analyze cte definitions and substitute child plan with analyzed cte definitions.
+   * Analyze WITH nodes and substitute child plan with CTE definitions.
    */
   object CTESubstitution extends Rule[LogicalPlan] {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: How about making a new file for this rule? cc: @gatorsmile

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -212,31 +212,96 @@ class Analyzer(
...
-      case With(child, relations) =>
-        // substitute CTE expressions right-to-left to resolve references to previous CTEs:
-        // with a as (select * from t), b as (select * from a) select * from b
-        relations.foldRight(child) {
-          case ((cteName, ctePlan), currentPlan) =>
-            substituteCTE(currentPlan, cteName, ctePlan)
-        }
-      case other => other
-    }
+    def apply(plan: LogicalPlan): LogicalPlan = traverse(plan, false)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I feel the current logic is a little hard to understand because of the legacy mode.
I think the existing logic for WITH is pretty simple, so how about keeping the existing one and adding a new logic for subqueries like this?
```
  object CTESubstitution extends Rule[LogicalPlan] {
    def apply(plan: LogicalPlan): LogicalPlan = if (!SQLConf.get.legacyCTESubstitutionEnabled) {
      traverseAndsubstituteCTE(plan)
    } else {
      legacyTraverseAndsubstituteCTE(plan)
    }
```

File: sql/core/src/test/resources/sql-tests/inputs/cte.sql [hunk 4]
```diff
@@ -28,6 +28,177 @@ FROM   CTE1 t1
...
+);
+
+-- CTE in subquery expression
+SELECT (
+  WITH t AS (SELECT 1)
+  SELECT * FROM t
+);
+
+-- CTE in CTE definition shadows outer (legacy substitution)
+SET spark.sql.legacy.cte.substitution.enabled=true;
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/cte.sql]: Could we split these tests into two separate files: the legacy one and non-legacy one?

**Review by maropu — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 1 above)
Comment (peter-toth) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: ok, done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Thanks @maropu for the review! I changed as you suggested.

File: sql/core/src/test/resources/sql-tests/inputs/cte.sql (hunk 4 above)
Comment (peter-toth) [reply sql/core/src/test/resources/sql-tests/inputs/cte.sql]: ok, done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: ok moved

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+   *   definition.
+   *   For example the following query should return 2:
+   *   WITH t AS (SELECT 1)
+   *   SELECT (
+   *     WITH t AS (SELECT 2)
+   *     SELECT * FROM t
+   *   )
+   * @param plan  the plan to be traversed
+   * @param inTraverse whether the current traverse is called from another traverse, only in this
+   *                   can name collision occur
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]:
```suggestion
   *                   case name collision can occur
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        }
+
+        relations.foldRight(traversedChild) {
+          case ((cteName, ctePlan), currentPlan) =>
+            substituteCTE(currentPlan, cteName, traverseAndSubstituteCTE(ctePlan, true))
+        }
+
+      case o if inTraverse =>
+        o.transformExpressions {
+          case e: SubqueryExpression => e.withNewPlan(traverseAndSubstituteCTE(e.plan, inTraverse))
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: `inTraverse` can be replaced with `true`, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        }
+    }
+
+  // A CTE definition might not be used at all so ctePlan is call by name.
+  private def substituteCTE(
+      plan: LogicalPlan,
+      cteName: String,
+      ctePlan: => LogicalPlan): LogicalPlan = {
+    lazy val lazyCTEPlan = ctePlan
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Thanks @mgaido91! It is not needed indeed.

[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: I remember now why I used lazy here. A CTE definition can be used multiple times in `WITH` but the call by name parameter (`ctePlan  = traverseAndSubstituteCTE(...)`) should be executed only once.
But now I believe it is better to use lazy outside of `substituteCTE` than inside, please review my commit https://github.com/apache/spark/pull/24831/commits/7d6910599134dc992bc841dfd3eff6f46d36647b.

File: sql/core/src/test/resources/sql-tests/inputs/cte-legacy.sql [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+WITH t AS (SELECT 1)
+SELECT (
+  WITH t AS (SELECT 2)
+  SELECT (
+    WITH t AS (SELECT 3)
+    SELECT * FROM t
+  )
+);
+
+SET spark.sql.legacy.cte.substitution.enabled=false;
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/inputs/cte-legacy.sql]: you don't need this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+ * Analyze WITH nodes and substitute child plan with CTE definitions.
+ */
+object CTESubstitution extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = if (SQLConf.get.legacyCTESubstitutionEnabled) {
+    legacyTraverseAndsubstituteCTE(plan)
+  } else {
+    traverseAndSubstituteCTE(plan, false)
+  }
+
+  def legacyTraverseAndsubstituteCTE(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: private?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+          case e: SubqueryExpression => e.withNewPlan(traverseAndSubstituteCTE(e.plan, true))
+        }
+
+        relations.foldRight(traversedChild) {
+          case ((cteName, ctePlan), currentPlan) =>
+            lazy val substitutedCTEPlan = traverseAndSubstituteCTE(ctePlan, true)
+            substituteCTE(currentPlan, cteName, substitutedCTEPlan)
+        }
+
+      case o if inTraverse =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Why we need this flag? Some tests would fail if we have not?

**Comment by maropu** [comment]
@peter-toth Can we split this pr into two parts cleanly: 1. the sub-query support and 2. the behaviour change? Also, I think we need to update the migration guide for the part 2.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 10 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: No it would not, but I wanted to do CTE substitution in the current plan only (not in the subqueries) if it is safe. (CTE substitution will run for subqueries later anyway.)

**Comment by peter-toth** [comment]
> @peter-toth Can we split this pr into two parts cleanly: 1. the sub-query support and 2. the behaviour change? Also, I think we need to update the migration guide for the part 2. cc: @gatorsmile

All right, I dropped the changes that relate to order of substitution and will do it in another ticket. What remained here is just the `WITH` support in subqueries.

**Comment by maropu** [comment]
Yea, thanks!

**Comment by peter-toth** [comment]
> Yea, thanks!

Very welcome.
I moved substitution order related changes to  https://github.com/apache/spark/pull/25029, but it is still WIP and this one should be reviewed first.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.catalyst.expressions.SubqueryExpression
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, With}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Analyze WITH nodes and substitute child plan with CTE definitions.
+ */
+object CTESubstitution extends Rule[LogicalPlan] {
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: can we avoid moving the class, in order to keep the diff smaller?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 12]
```diff
@@ -412,7 +412,7 @@ queryPrimary
     | fromStatement                                                         #fromStmt
     | TABLE multipartIdentifier                                             #table
     | inlineTable                                                           #inlineTableDefault1
-    | '(' queryNoWith  ')'                                                  #subquery
```
Comment (mgaido91) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: can we also remove `queryNoWith` since it has only 1 usage?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 13]
```diff
@@ -211,37 +211,6 @@ class Analyzer(
...
-            substituteCTE(currentPlan, cteName, ctePlan)
-        }
-      case other => other
-    }
-
-    def substituteCTE(plan: LogicalPlan, cteName: String, ctePlan: LogicalPlan): LogicalPlan = {
-      plan resolveOperatorsUp {
-        case UnresolvedRelation(Seq(table)) if resolver(cteName, table) =>
-          ctePlan
-        case u: UnresolvedRelation =>
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: why did you remove this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 11 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: The idea of moving the rule to a separate file came from here: https://github.com/apache/spark/pull/24831#discussion_r298864911
, but I think you are right @mgaido because since that we cut the scope and split the PR. Maybe the other part (https://github.com/apache/spark/pull/25029) could extract the rule to a separate file as that one makes the rule a bit more complicated. Does that work for you @maropu?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 12 above)
Comment (peter-toth) [reply sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Thanks good idea!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 13 above)
Comment (peter-toth) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I don't think this like does anything nor `UnresolvedRelation` can have an expression so I thought it is safe and good idea to remove it. Please correct me if I'm wrong.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: yes, I think you're right, I was just curious about the reason of this change

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 14]
```diff
@@ -226,20 +226,20 @@ class Analyzer(
       case other => other
     }
 
-    def substituteCTE(plan: LogicalPlan, cteName: String, ctePlan: LogicalPlan): LogicalPlan = {
+    private def substituteCTE(
+        plan: LogicalPlan,
+        cteName: String,
+        ctePlan: LogicalPlan): LogicalPlan =
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: may you please keep the  `{`. I think in general we are trying to always use them in order to help readability.

**Review by mgaido91 — COMMENTED** [review]
only a style comment, otherwise LGTM, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala (hunk 11 above)
Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala]: Yes, that was what I meant, we can move the rule in the other PR which refactors it more thoroughly.

File: sql/core/src/test/resources/sql-tests/results/cte.sql.out [hunk 15]
```diff
@@ -98,16 +98,202 @@ struct<x:int>
...
+1
+
+
+-- !query 20
+WITH r AS (SELECT * FROM r)
+SELECT * FROM r
+-- !query 20 schema
+struct<>
+-- !query 20 output
+org.apache.spark.sql.AnalysisException
```
Comment (mgaido91) [inline sql/core/src/test/resources/sql-tests/results/cte.sql.out]: Just a question, is this going to be addressed in the PR which allows recursive subqueries or is this an invalid query?

Comment (peter-toth) [reply sql/core/src/test/resources/sql-tests/results/cte.sql.out]: I have a WIP PR open https://github.com/apache/spark/pull/23531 that would add support for recursive queries (and subqueries and subquery expressions too), but using recursive reference in a subquery is not allowed according to the SQL standard. 

But, this PR should be accepted first then could come https://github.com/apache/spark/pull/25029 and https://github.com/apache/spark/pull/23531

Actually I think I'm removing the test `WITH r AS (SELECT * FROM r) SELECT * FROM r;` because there is already a similar one in `cte.sql` and moving the `WITH r AS (SELECT (SELECT * FROM r)) SELECT * FROM r;` next to the existing one.

File: sql/core/src/test/resources/sql-tests/results/cte.sql.out [hunk 16]
```diff
@@ -88,26 +98,192 @@ struct<c1:int,c2:int>
...
+  t2 AS (
+    WITH t AS (SELECT 2)
+    SELECT * FROM t
+  )
+SELECT * FROM t2
+-- !query 12 schema
+struct<1:int>
+-- !query 12 output
+1
+
```
Comment (maropu) [inline sql/core/src/test/resources/sql-tests/results/cte.sql.out]: This result is different from the pg one;
```

postgres=# WITH
postgres-#   t AS (SELECT 1),
postgres-#   t2 AS (
postgres(#     WITH t AS (SELECT 2)
postgres(#     SELECT * FROM t
postgres(#   )
postgres-# SELECT * FROM t2;
 ?column? 
----------
        2
(1 row)
```
This will be address in the following  #25029?

Comment (dongjoon-hyun) [reply sql/core/src/test/resources/sql-tests/results/cte.sql.out]: I also agree that this is inevitable in this PR. (cc @gatorsmile ).

Comment (peter-toth) [reply sql/core/src/test/resources/sql-tests/results/cte.sql.out]: Yes, after https://github.com/apache/spark/pull/25029 it will return `2` (https://github.com/apache/spark/pull/25029/files#diff-fc515a5db268d29b08b80f5eb8202026R145)

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @peter-toth , @maropu , @mgaido91 .
The original PR is split into two (this and #25029) according to the review comment.
This is a new feature at Spark 3.0.0 and will be consistent with PostgreSQL soon.
Merged to master to move forward.

**Comment by peter-toth** [comment]
Thanks @dongjoon-hyun, @maropu, @mgaido91 for the review! I will prepare https://github.com/apache/spark/pull/25029 for review soon.
